portalnet/src/utp_controller.rs

```diff
@@ -15,33 +22,235 @@ use crate::discovery::UtpEnr;
 pub struct UtpController {
     pub inbound_utp_transfer_semaphore: Arc<Semaphore>,
     pub outbound_utp_transfer_semaphore: Arc<Semaphore>,
-    /// uTP socket.
     utp_socket: Arc<UtpSocket<UtpEnr>>,
+    metrics: OverlayMetricsReporter,
+}
+
+lazy_static! {
+    /// The default configuration to use for uTP connections.
+    pub static ref UTP_CONN_CFG: ConnectionConfig = ConnectionConfig { max_packet_size: 1024, ..Default::default()};
 }
 
 impl UtpController {
-    pub fn new(utp_transfer_limit: usize, utp_socket: Arc<UtpSocket<UtpEnr>>) -> Self {
+    pub fn new(
+        utp_transfer_limit: usize,
+        utp_socket: Arc<UtpSocket<UtpEnr>>,
+        metrics: OverlayMetricsReporter,
+    ) -> Self {
         Self {
             utp_socket,
             inbound_utp_transfer_semaphore: Arc::new(Semaphore::new(utp_transfer_limit)),
             outbound_utp_transfer_semaphore: Arc::new(Semaphore::new(utp_transfer_limit)),
+            metrics,
+        }
+    }
+
+    /// Connect with a peer given the connection id, and return the data received from the peer.
+    pub async fn connect_inbound_stream(
+        &self,
+        connection_id: u16,
+        peer: Enr,
+    ) -> anyhow::Result<Vec<u8>> {
+        self.metrics
+            .report_utp_active_inc(UtpDirectionLabel::Inbound);
+        let cid = utp_rs::cid::ConnectionId {
+            recv: connection_id,
+            send: connection_id.wrapping_add(1),
+            peer: UtpEnr(peer),
+        };
+        let stream = match self.connect_with_cid(cid.clone(), *UTP_CONN_CFG).await {
+            Ok(stream) => stream,
+            Err(err) => {
+                self.metrics.report_utp_outcome(
+                    UtpDirectionLabel::Inbound,
+                    UtpOutcomeLabel::FailedConnection,
+                );
+                debug!(
+                    %err,
+                    cid.send,
+                    cid.recv,
+                    peer = ?cid.peer.client(),
+                    "Unable to establish uTP conn based on Content response",
+                );
+                return Err(anyhow!("unable to establish utp conn"));
+            }
+        };
+
+        // receive_utp_content handles metrics reporting of successful & failed rx
+        match Self::receive_utp_content(stream, self.metrics.clone()).await {
+            Ok(data) => Ok(data),
+            Err(err) => {
+                debug!(%err, cid.send, cid.recv, peer = ?cid.peer.client(), "error reading data from uTP stream, while handling a FindContent request.");
+                Err(anyhow!("error reading data from uTP stream"))
+            }
+        }
+    }
+
+    /// Connect with a peer given the connection id, and transfer the data to the peer.
```

Comment:
nitpick: I would say it makes more sense and it's consistent if `connect_outbound_stream` returns `anyhow::Result<()>`. You can also do proper error logging one level up.